projectq/meta/_control.py

```diff
@@ -101,15 +111,49 @@ def __init__(self, engine, qubits):
             qubits = [qubits]
         self._qubits = qubits
 
+        # If the user inputs a single digit 0 or 1, extend the digit to all ctrl qubits
+        if ctrl_state == 0  or ctrl_state == '0':
+            self._state = str(ctrl_state)*len(qubits)
+        elif type(ctrl_state) is State:
+            if ctrl_state.value == -1:
+
+                self._state = '1'*len(qubits)
+            else:
+                self._state = '0' * len(qubits)
+        # If the user inputs an integer, convert it to binary bit string
+        elif type(ctrl_state) is int:
+            bit_length = len(self._qubits)
+            self._state = '{0:b}'.format(ctrl_state).zfill(bit_length)
+
+        # If the user inputs bit string, directly use it
+        elif type(ctrl_state) is str:
+            self._state = ctrl_state
+
+        else:
+            raise TypeError('Input must be a string, an integer or class State')
+        # Raise exceptions for wrong cases: invalid string length and number
+        assert len(self._state) == len(self._qubits), 'Control state has different length than control qubits'
+        assert set(self._state).issubset({'0','1'}), 'Control state has string other than 1 and 0'
+
+
     def __enter__(self):
         if len(self._qubits) > 0:
-            ce = ControlEngine(self._qubits)
+
+            with Compute(self.engine):
+                for i in range(len(self._state)):
+                    if self._state[i]=='0':
+                        X | self._qubits[i]
+
```

Comment:
# General comments

Thinking about it some more, I think we should take a different approach to add the required control qubits that are required to invert the state of the control qubits if the user requests it.

Currently, you pass the state of the qubits to the `ControlEngine` *and* call the `Compute/Uncompute` to make sure the control qubits are in the proper state. This is a bit redundant, because if you already apply some X gates on the control qubits, then why do you need to pass the state of the control qubits to the `ControlEngine` (the state information has already been taken care of at this point, right?)

Therefore, in order to make sure that if the backend supports it, we can pass the state of the control qubits *as-is*, I propose the following:
  - Remove the `Compute/Uncompute` section in the `Control` context manager
  - Add a new decomposition rule in `projectq/setups/decompositions/` called `controlstate.py` [1]
  - Modify all current backends `is_available()` methods to reject negative controls by default [2]
  - Modify the default setup for all known backends to add this decomposition rule by default [3]

# [1] Decomposition rule

Writing the decomposition rule should be fairly straightforward with something like this:

```python
from projectq.cengines import DecompositionRule
from projectq.meta import Control, get_control_count, Compute, Uncompute, has_negative_control
from projectq.ops import BasicGate, X

def _decompose_controlstate(cmd):
    """
    Decompose commands with control qubits in negative state (ie. control
    qubits with state '0' instead of '1')
    """
    with Compute(cmd.engine):
        for state, ctrl in zip(cmd.ctrl_state, cmd.control_qubits):
            if state == '0':
                X | ctrl

    # Resend the command with the `ctrl_state` cleared
    cmd.ctrl_state = '1' * len(cmd.ctrl_state)
    cmd.engine.receive([cmd])

    Uncompute(cmd.engine)

#: Decomposition rules
all_defined_decomposition_rules = [
    DecompositionRule(BasicGate, _decompose_controlstate, has_negative_control)
]
```

and add this function at the bottom of `projectq/meta/_control.py`:

```python
def has_negative_control(cmd):
    return get_control_count(cmd) > 0 and '0' in cmd.ctrl_state
```


# [2] Backends `is_available`

For example in `projectq/backends/_sim/simulator.py` add this to the `is_available` method:

```python
    def is_available(self, cmd):
        # ...
        if has_negative_control(cmd):
                return False
        # ...
```

and same thing for all the other backends for now. When we figure out how to pass the control state information to their respective APIs, we can then lift this.

# [3] Setups default

Check that all the `get_engine_list()` functions in all of the files in `projectq/setups` automatically add this decomposition rule by default. Note that in some cases the rule will be added automatically if an `AutoReplacer` engine is present that includes all decomposition rules like in the case for the `default` setup (so nothing to do in this case):

```python
def get_engine_list():
    rule_set = DecompositionRuleSet(modules=[projectq.setups.decompositions]) # contains *all* decomposition rules
    return [TagRemover(),
            LocalOptimizer(10),
            AutoReplacer(rule_set),
            TagRemover(),
            LocalOptimizer(10)]
```

Also make sure that a `LocalOptimizer` engine is always present *after* the `AutoReplacer` that *potentially* will use the decomposition rule above to make sure that if multiple commands with the same control qubits pass one after the other we cancel the `X` gates in between.